python/ray/util/client/dataclient.py

```diff
@@ -60,48 +60,111 @@ def _start_datathread(self) -> threading.Thread:
+                    return
+                except grpc.RpcError as e:
+                    reconnecting = self._process_rpc_error(e)
+                    if not reconnecting:
+                        break
+        finally:
+            logger.info("Shutting down data channel")
+            self._shutdown()
+
+    def _process_response(self, response: Any) -> None:
+        """
+        Process responses from the data servicer.
+        """
+        if response.req_id == 0:
+            # This is not being waited for.
+            logger.debug(f"Got unawaited response {response}")
+            return
+        # Acknowledge to server that response has been received -- server
+        # can clear cache entry
+        self._acknowledge(response.req_id)
+        if response.req_id in self.asyncio_waiting_data:
+            # Async response. Check for callback and update outstanding
+            # requests.
+            callback = self.asyncio_waiting_data.pop(response.req_id)
+            try:
+                if callback:
+                    callback(response)
+            except Exception:
+                logger.exception("Callback error:")
+            if response.req_id in self.outstanding_requests:
+                del self.outstanding_requests[response.req_id]
+        else:
+            # Blocking response. Populate ready data and wake up waiting
+            # threads.
             with self.cv:
-                self._in_shutdown = True
+                self.ready_data[response.req_id] = response
                 self.cv.notify_all()
-            if e.code() == grpc.StatusCode.CANCELLED:
-                # Gracefully shutting down
-                logger.info("Cancelling data channel")
-            elif e.code() in (grpc.StatusCode.UNAVAILABLE,
-                              grpc.StatusCode.RESOURCE_EXHAUSTED):
-                # TODO(barakmich): The server may have
-                # dropped. In theory, we can retry, as per
-                # https://grpc.github.io/grpc/core/md_doc_statuscodes.html but
-                # in practice we may need to think about the correct semantics
-                # here.
-                logger.info("Server disconnected from data channel")
-            else:
-                logger.exception(
-                    "Got Error from data channel -- shutting down:")
+
+    def _acknowledge(self, req_id: int):
+        self.request_queue.put(
+            ray_client_pb2.DataRequest(
+                acknowledge=ray_client_pb2.AcknowledgeRequest(req_id=req_id)))
+
+    def _process_rpc_error(self, e: grpc.RpcError) -> bool:
+        """
+        Processes RPC errors that occur while reading from data stream.
+        Returns True if the error can be recovered from, False otherwise.
+        """
+        if not self.client_worker._can_reconnect(e):
+            logger.info("Unrecoverable error in data channel.")
+            return False
+
+        # Ping the server to see if the current channel is reuseable
```

Comment:
Can we break out the retry logic into another function with a clear name, e.g. `reset_connection()`? It is a bit unexpected for have `_process_rpc_error` running reconnection underneath.